pkg/controller/worker/reconciler.go

```diff
@@ -84,91 +84,166 @@ func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, err
 		return reconcile.Result{}, err
 	}
 
-	// Deletion flow
-	if worker.DeletionTimestamp != nil {
-		hasFinalizer, err := extensionscontroller.HasFinalizer(worker, FinalizerName)
-		if err != nil {
-			r.logger.Error(err, "Could not instantiate finalizer deletion")
-			return reconcile.Result{}, err
-		}
-		if !hasFinalizer {
-			r.logger.Info("Deleting worker causes a no-op as there is no finalizer.", "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
-			return reconcile.Result{}, nil
-		}
+	operationType := gardencorev1beta1helper.ComputeOperationType(worker.ObjectMeta, worker.Status.LastOperation)
```

Comment:
You could hand over the operationType to the delete, restore and reconcile functions? (instead of re-calculating them)